cli/admin/src/main/java/io/pravega/cli/admin/dataRecovery/RecoverFromStorageCommand.java

```diff
@@ -0,0 +1,595 @@
+        @Cleanup
+        val dataLogFactory = new BookKeeperLogFactory(bkConfig, zkClient, executorService);
+        output("Container Count = %d", this.containerCount);
+        dataLogFactory.initialize();
+        output("Started ZK Client at %s.", getServiceConfig().getZkURL());
+        output("Starting recovery...");
+        // STEP 1: Get all the segment chunks related to the main Segment in order.
+        File[] allFiles = new File(tableSegmentDataChunksPath).listFiles();
+
+        int containerId = NumberUtils.isNumber(container) ? Integer.parseInt(container) : 0;
+        int endContainer = containerId + 1;
+        if (container.equalsIgnoreCase(ALL_CONTAINERS)) {
+            containerId = 0;
+            endContainer = getServiceConfig().getContainerCount();
+        }
+
+        while ( containerId < endContainer ) {
+            recoverFromStorage(containerId, allFiles, context, dataLogFactory);
+            containerId++;
+        }
+    }
+
+    /**
+     * Performs actual recovery procedure.
+     * 1. Spins up a DebugSegmentContainer.
+     * 2. Parses raw chunks into operations.
+     * 3. Writes the Operations to the spinned up container into respective storage_metata and container_metadata tables.
+     * 4. Validates all chunks are written.
+     * 5. Reconciles segments between storage_metdadata and container_metadata with storage_metadata as truth.
+     * 6. Overrides DurabaleDataLog Metadata with the highest epoch seen.
+     * @param containerId Id of the container to be created.
+     * @param allFiles all chunk files to be parsed.
+     * @param context Context holding config.
+     * @param dataLogFactory Factory of DurabeleDataLog to be created.
+     * @throws Exception
+     */
+    private void recoverFromStorage(int containerId, File[] allFiles, Context context, DurableDataLogFactory dataLogFactory ) throws Exception {
+        File[] potentialFiles = getContainerMetadataChunkFiles(allFiles, containerId);
+        deleteJournalsOf(containerId);
+        //create debug segment container
+        DebugStreamSegmentContainer debugStreamSegmentContainer = createDebugSegmentContainer(context, containerId, dataLogFactory);
+        output("-----------------DebugSegment container %d initialized--------------", containerId);
+        // segregate the metadata and storage metadata chunks
+        Map<String, List<File>> metadataSegments = segregateMetadataSegments(potentialFiles);
+        Preconditions.checkState(metadataSegments.size() == 2, "Only MetadataSegment and Storage MetadataSegment chunks should be present");
+        List<File> storageChunks = null;
+        boolean firstRun = true;
+        int attempts = 0;
+
+        ChunkValidator chunkValidator = new ChunkValidator(debugStreamSegmentContainer);
+        Map<Integer, Long> containersToEpoch = new HashMap<>();
+        while (firstRun || (!chunkValidator.validate() /*&& attempts < RETRY_ATTEMPT*/)) {
```

Comment:
`/*&& attempts < RETRY_ATTEMPT*/ ` is this intentional ?

Also add a comment regarding use of firstRun. It is not obvious why we need to track that

Reply:
RETRY_ATTEMPT is intentional for now, to let it run, so that we can see some diff of the validation in the logs. RETRY_ATTEMPTS will be enabled in subsequent improvements to this command.
firstRun was basicaly simulating do..while.. Have removed it and changed it to do..while.